packages/core/src/shared/awsClientBuilderV3.ts

```diff
@@ -78,19 +88,37 @@ export class AWSClientBuilderV3 {
         return shim
     }
 
-    public async createAwsService<C extends AwsClient>(
-        type: AwsClientConstructor<C>,
-        options?: Partial<AwsClientOptions>,
-        region?: string,
-        userAgent: boolean = true,
-        keepAlive: boolean = true,
-        settings?: DevSettings
-    ): Promise<C> {
+    private keyAwsService<C extends AwsClient>(serviceOptions: AwsServiceOptions<C>): string {
+        // Serializing certain objects in the args allows us to detect when nested objects change (ex. new retry strategy, endpoints)
+        return [
+            String(serviceOptions.serviceClient),
+            JSON.stringify(serviceOptions.clientOptions),
+            serviceOptions.region,
+            serviceOptions.userAgent ? '1' : '0',
+            serviceOptions.settings ? JSON.stringify(serviceOptions.settings.get('endpoints', {})) : '',
+        ].join(':')
+    }
+
+    public async getAwsService<C extends AwsClient>(serviceOptions: AwsServiceOptions<C>): Promise<C> {
+        const key = this.keyAwsService(serviceOptions)
+        const cached = this.serviceCache.get(key)
+        if (cached) {
+            return cached as C
+        }
+
+        const service = await this.createAwsService(serviceOptions)
+        this.serviceCache.set(key, service)
+        return service as C
+    }
+
+    public async createAwsService<C extends AwsClient>(serviceOptions: AwsServiceOptions<C>): Promise<C> {
```

Comment:
should the docstring call out the fact that the client will be re-used?

important for callers not to "modify" the client, except in specific "safe" ways...